src/index.js

```diff
@@ -78,22 +78,30 @@ router.get('/', ctx => {
 app.use(koaStatic(path.join(__dirname, '../static/')));
 app.use(router.routes(), router.allowedMethods());
 
-const apolloServer = new ApolloServer({
+export const apolloServer = new ApolloServer({
   schema,
   introspection: true, // Allow introspection in production as well
   playground: true,
-  context: ({ ctx }) => ({
-    loaders: new DataLoaders(), // new loaders per request
-    user: ctx.state.user,
-
-    // userId-appId pair
-    //
-    userId:
-      ctx.appId === 'WEBSITE' || ctx.appId === 'DEVELOPMENT_FRONTEND'
-        ? (ctx.state.user || {}).id
-        : ctx.query.userId,
-    appId: ctx.appId,
-  }),
+  context: async ({ ctx }) => {
+    let currentUser = ctx.state.user || {};
+    if (ctx.appId) {
+      const { user } = await createOrUpdateUser({
+        appUserId: ctx.query.userId,
+        appId: ctx.appId,
+      });
+      currentUser = { ...currentUser, ...user };
```

Comment:
1. I cannot think of a case that `appId` is missing -- it should always be populated by `checkHeaders()`. If this is expected, maybe we can simplify the logic here by rejecting empty `ctx.appId` early.
2. Since we are invoking `createOrUpdateUser`, which will invoke database anyway, I think we can consider
    - remove DB read operation in `passport.deserializeUser` in `auth.js` -- we can just return `userId` or `{passportUserId: userId}` in `passport.deserializeUser`
    - no need to merge `ctx.state.user` with GraphQL context's user